### ISSUE TITLE
Reconsider macOS config locations

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#5179]
This is mostly a continuation of #383. Various things come to mind:

# autoconfig path

- Qt gives us `~/Library/Preferences` for a macOS config path. However, looking at the [macOS documentation](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW1) that seems like a bad idea (emphasis mine):

> This directory contains app-specific preference files. **You should not create files in this directory yourself.** Instead, use the NSUserDefaults class or CFPreferences API to get and set preference values for your app.

The "application support" folder seems to be more appropriate:

> Use this directory to store all app data files except those associated with the user’s documents. For example, you might use this directory to store app-created data files, **configuration files**, templates, or other fixed or modifiable resources that are managed by the app.

This seems like a bug in Qt's QStandardPaths to me - the only related discussion I could find is: [[QTBUG-40737] OSX: be consistent in using either ~/.config/QtProject or ~/Library/Preferences for ini files - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-40737)

# config path

We currently use `~/.qutebrowser/config` for the "human-writable" part of configs. Some users would prefer qutebrowser to use an XDG-like `~/.config/qutebrowser` instead. Fully supporting XDG probably is a bad idea (for example, the cache folder should stay at `~/Library/Caches` which is excluded from Time Machine backups by default). However, I guess supporting `$XDG_CONFIG_HOME` (defaulting to `~/.config`) wouldn't be a too bad idea.

Another question is how we handle the migration - some possibilities:

- Move `~/.qutebrowser` to `~/.config/qutebrowser` automatically
- Read `~/.qutebrowser` if it exists (and ignore `~/.config/qutebrowser`), but default to `~/.config/qutebrowser` if not
- Read both - for `config.py` that *should* be possible, for quickmarks/bookmarks probably too? Might not be worth the effort, though. On the positive side, it'd allow us to also consider `XDG_CONFIG_DIRS` (e.g. `/etc/xdg/qutebrowser`) on Linux.

**Comment by The-Compiler** [comment]
As a data point, some XDG folder contents on macOS from [this Reddit post](https://www.reddit.com/r/commandline/comments/e935c1/what_are_some_clitui_applications_with_vilike/fdm2os4?utm_source=share&utm_medium=web2x):

```
~/.local/share/next # Next browser
~/.local/share/heroku
~/.local/share/bower
~/.local/share/nvim # neovim
# created by Meld merge, GTK app w/ unofficial macOS builds, so not the best example
~/.local/share/recently-used.xbel
~/.local/share/applications # created by WineBottler
~/.local/share/mime # created by WineBottler
~/.local/share/desktop-directories # created by WineBottler
# created by Meld merge, populated by recents referenced in `recently-used.xbel`
~/.local/share/meld
~/.local/share/karabiner # created by Karabiner Elements, macOS-exclusive app
```

```
~/.config/alacritty
~/.config/glade.conf
~/.config/heroku
~/.config/configstore
~/.config/nvim
~/.config/cmus
~/.config/gtk-3.0
~/.config/yarn
~/.config/inkscape
~/.config/htop
~/.config/gtk-2.0
~/.config/rtv
~/.config/mpv
~/.config/minidlna
~/.config/menus
~/.config/kitty
~/.config/fontforge
~/.config/fish
~/.config/hub
~/.config/git
~/.config/karabiner
```

```
~/.cache/proselint
~/.cache/ruby-install
~/.cache/gegl-0.2 # GIMP?
~/.cache/inkscape
~/.cache/fontconfig
~/.cache/ctrlp
~/.cache/youtube-dl
~/.cache/neocomplcache
~/.cache/cabal-helper
~/.cache/babl # GIMP?
```

So using `~/.config` but not the rest probably wouldn't be too bad (though some of those examples might put data files in `~/.config` as well?).

**Comment by marco-silva0000** [comment]
is there a way to manually set where the file should be?

**Comment by The-Compiler** [comment]
@marco-silva0000 Which file? There's `--config-py`, other than that, no.

**Comment by Granitosaurus** [comment]
There seems to be no way edit flags of isntalled mac applications right? 
currently I'm using applescript:
```
do shell script "open -a qutebrowser --args -B ~/.config/qutebrowser -C ~/.config/qutebrowser/config.py"
```
To replicate linux configuration. Then save it as `.app`. However now there are two entry points to qutebrowser that polute all of the app namespaces :(

Is ther alternative way to set application flags in MacOs?

**Comment by The-Compiler** [comment]
I'm afraid this isn't going to make it into v2.0.0.

**Comment by keegancsmith** [comment]
A simple solution which avoids the need to migrate data/steps from users is to support multiple locations. On startup you pick the first location you find. For example this is what emacs did (which added support for XDG using this approach) https://www.gnu.org/software/emacs/manual/html_node/emacs/Find-Init.html#Find-Init I'm not familiar with the qutebrowser codebase, but would that be a reasonable solution?

**Comment by ftpd** [comment]
I second @keegancsmith's idea. This way all 'new' users will be XDG-compliant out of the box and anyone who already uses qutebrowser could choose to move their config or not.

**Comment by gerrywastaken** [comment]
* Move ~/.qutebrowser to ~/.config/qutebrowser automatically

I like this option, but you would have to add an alert when doing so. Many sync their config files from a git repo into specific locations and so you would move it but then then next time they resynced it would just be put back.

**Comment by wongjiahau** [comment]
At the mean time, is there any workaround to load the config from `~/.config/qutebrowser` on macOS?

**Comment by nshern** [comment]
Is there any solution for this?

**Comment by The-Compiler** [comment]
Nothing I'm aware of, otherwise it'd be in this issue.

**Comment by The-Compiler** [comment]
I don't think so - happy to assign this to you if you want to take a look. Note that a contribution will need to provide answers for various questions raised above, especially around how to best migrate things to the new location. It will also need to consider various corner cases, e.g. what to do if both locations exist.

**Comment by mojadem** [comment]
Awesome, thanks for the quick reply. The solution I propose mirrors the one @keegancsmith suggested: support both locations. I think the [Wezterm documentation](https://wezfurlong.org/wezterm/config/files.html#configuration-files) goes over the corresponding logic quite well. Essentially, there will be an order of locations qutebrowser checks for a config (`$XDG_CONFIG_HOME/qutebrowser`, then `~/.config/qutebrowser`, then `~/.qutebrowser`). This way, existing configs can remain where they are. I don't think we need to do any merging of configs or automatic moving.

**Comment by The-Compiler** [comment]
Unassigning you again as I'm assuming you're not working on this anymore?

**Comment by Noaemien** [comment]
I'm working on the solution proposed by @mojadem which is a priority order for config folder location.
- `$XDG_CONFIG_HOME/qutebrowser` defaulting to `~/.config/qutebrowser`
- `~/.qutebrowser`
Allowing people with already existing configs to keep them in place.

Tests need to be adapted for the CI to pass.
A few tests can be added as well to tests all the different possible folder configurations.

**Comment by davelens** [comment]
+1 for @Noaemien's PR. Fresh user here, sadly still on macos, came looking specifically for instructions on configuring it to `~/.config/qutebrowser`. Glad to see this seems to be possible in the near future.

**Comment by turkishmaid** [comment]
Please review and merge PR. Waiting for this so much...

🔥 When started like `qutebrowser --basedir "/Users/sara/.config/qutebrowser"`, it puts `cache/`, `config/`, and `data/` folders under `~/.config/qutebrowser`, which is also a no-no...

Can't use use anything not respecting XDG shell vars 🤷‍♀️

Workaround (in `.zshrc`) for the time:

```sh
# XDG Base Directory Specification
export XDG_CONFIG_HOME="$HOME/.config"

qq() {
    # Run the browser with a junk-quarantine basedir
    /Applications/qutebrowser.app/Contents/MacOS/qutebrowser \
        --config-py "$XDG_CONFIG_HOME/qutebrowser/config.py" \
        --basedir "$XDG_CONFIG_HOME/junk/qutebrowser" "$@"
}
```
